### PR TITLE
Add automatic best model loading to Trainer

### LINKED ISSUE #6901: Relaxing `PreTrainedModel` requirement in _save
# 🚀 Feature request
It's great to see that `Trainer` is becoming flexible. Each functions seems to be more self contained now making inheritance easier. I've experimented with many custom models. For instance, 
```
class Model(nn.Module):
    def __init__(self, ..):
        self.encoder = AutoModel.from_pretrained(..)
        self.custom_modules = ..
    def forward(self, **kwargs):
        output = self.encoder(**kwargs)
        # some custom operations
```
Many users are required to create custom models if they just don't want simple `SequenceClassification` head. In all cases, I have to override `_save` method because of [this line](https://github.com/huggingface/transformers/blob/d822ab636b6a14ed50f7bca0797c1de42c19de61/src/transformers/trainer.py#L1097) which explicitly puts a restriction on `Trainer` to be used with models that inherit from `PreTrainedModel`. It would be good to relax this requirement and give a warning about not using `PreTrainedModel` instead.

## Your contribution
I'll open a PR if I get approval.

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR cleans up a bit the part that saves the training state inside `Trainer` and adds an API that can track which was the best model during any of the evaluation phases to load it back at the end.

When fine-tuning a model on a dataset that can easily overfit the model, it's quite common to have the last model not be the best one (in terms of metrics). This PR adds a `TrainingArgument` named `load_best_model_at_end` that triggers the following behavior:
- `save_steps` gets ignored and the model is saved every time there is an evaluation (determined by `evaluation_strategy` and `eval_steps`)
- It keeps track in a `TrainerState` of when the best model was encountered (that state is saved along the checkpoints so it can work with resuming a training)
- The best model is determined by the new `TrainingArgument`s `metric_for_best_model` (defaults to the loss) and `greater_is_better` (default to False for the loss, True otherwise).
- The best model is loaded once the training is finished.

In passing I've added some tests of the saving API in Trainer and made sure it can handle both `PreTrainedModel` and regular `nn.Module` (a feature asked in #6901). Both are now tested in the CI, as is the new API.

Fixes #6901

Those newly introduced arguments and APIs can then be leveraged to have early stopping supported in `Trainer`.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1215,6 +1264,14 @@ def _sorted_checkpoints(self, checkpoint_prefix=PREFIX_CHECKPOINT_DIR, use_mtime
...
         checkpoints_sorted = sorted(ordering_and_checkpoint_path)
         checkpoints_sorted = [checkpoint[1] for checkpoint in checkpoints_sorted]
+        # Make sure we don't delete the best model.
+        if self.state.best_model_checkpoint is not None:
+            best_model_index = checkpoints_sorted.index(self.state.best_model_checkpoint)
+            checkpoints_sorted[best_model_index], checkpoints_sorted[best_model_index][-1] = (
+                checkpoints_sorted[-1],
+                checkpoints_sorted[best_model_index],
+            )
+            print(checkpoints_sorted)
```
Comment (julien-c) [inline src/transformers/trainer.py:1274]: bogus print

**Review by julien-c — APPROVED** [review]

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:1274]: Oopsie, leftover from debug.

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -145,6 +145,28 @@ class TrainingArguments:
...
+            loss).
+
+            If you set this value, :obj:`greater_is_better` will defaut to :obj:`True`. Don't forget to set it to
+            :obj:`False` if your metric is better when lower.
+        greater_is_better (:obj:`bool`, `optional`)
+            Use in conjunction with :obj:`load_best_model_at_end` and :obj:`metric_for_best_model` to specify if better
+            models should have a greater metric or not. Will default to:
+
+            - :obj:`True` if :obj:`metric_for_best_model` is set to a value that isn't :obj:`"loss"` or
+              :obj:`"eval_loss"`.
```
Comment (LysandreJik) [inline src/transformers/training_args.py:168]: Very clear doc!

**Review by LysandreJik — APPROVED** [review]
Nice! Cool that we allow other `nn.Module`s than `PreTrainedModel`s now!

**Comment by PhilipMay** [comment]
IMO this closes #4186

**Comment by PhilipMay** [comment]
@sgugger how does this work together with `save_total_limit` ? If it is set might it happen that the best model gets deleted?

**Comment by sgugger** [comment]
The best model is not deleted with `save_total_limit`. It is always put at the top of the list after sorting the chceckpoints.
